fairlearn/post_processing/roc_curve_based_post_processing.py

```diff
@@ -0,0 +1,378 @@
+        # negative/positive samples.
+        x, y = count[0] / n_negative, count[1] / n_positive
+        threshold = (threshold + scores[i]) / 2
+        operation = ThresholdOperation('>', threshold)
+
+        if flip and x > y:
+            x, y = 1 - x, 1 - y
+            operation = ThresholdOperation('<', threshold)
+        x_list.append(x)
+        y_list.append(y)
+        operation_list.append(operation)
+        
+    return pd.DataFrame({'x': x_list, 'y': y_list, 'operation': operation_list}).sort_values(by=['x', 'y'])
+
+def _get_scores_labels_and_counts(data):
+    data_sorted = data.sort_values(by=SCORE_KEY, ascending=False)
+
+    scores = list(data_sorted[SCORE_KEY])
+    labels = list(data_sorted[LABEL_KEY])
+
+    n, n_positive, n_negative = _get_counts(labels)
+
+    return scores, labels, n, n_positive, n_negative
+
+def _get_counts(labels):
+    n = len(labels)
+    n_positive = sum(labels)
+    n_negative = n - n_positive
+    return n, n_positive, n_negative
+
+def _get_roc(data, x_grid, attribute, flip=True, debug=False):
+    """Get ROC curve's convex hull based on data columns 'score' and 'label'
+    Scores represent output values from the model.
+    """
+    roc_sorted = _calculate_roc_points(data, attribute, flip)
+    roc_selected = _filter_points_to_get_convex_hull(roc_sorted)
+    roc_convex_hull = pd.DataFrame(roc_selected)[['x', 'y', 'operation']]
+    return roc_convex_hull
+
+def _filter_points_to_get_convex_hull(roc_sorted):
+    selected = []
+    for r2 in roc_sorted.itertuples():
+        while len(selected) >= 2:
+            r1 = selected[-1]
+            r0 = selected[-2]
+            # Calculate the y value at r2's x if the slope between r0 and r1 is continued.
+            # If that y value is not larger than r2's actual y value we know that r1 lies
+            # below the line between r0 and r2 and can be dropped since we can reach all points
+            # between r0 and r2 through interpolation.
+            if (r1.y - r0.y) * (r2.x - r0.x) <= (r2.y - r0.y) * (r1.x - r0.x):
+                selected.pop()
+            else:
+                break
+        selected.append(r2)
+    return selected
+        
+def _sanity_check_and_group_data(attributes, labels, scores):
+    if len(attributes) == 0 or len(labels) == 0 or len(scores) == 0:
+        raise ValueError(EMPTY_INPUT_ERROR_MESSAGE)
+
+    if len(attributes) != len(labels) or len(attributes) != len(scores):
+        raise ValueError(DIFFERENT_INPUT_LENGTH_ERROR_MESSAGE)
+    data = pd.DataFrame({ATTRIBUTE_KEY: attributes, SCORE_KEY: scores, LABEL_KEY: labels})
+
+    return data.groupby(ATTRIBUTE_KEY)
+    
+
+def roc_curve_based_post_processing_demographic_parity(attributes, labels, scores, gridsize=1000, flip=True, debug=False):
+    """ TODO add description
+    :param attributes: the protected attributes
+    :type attributes: list
+    :param labels: the labels of the dataset
```

Comment:
Don't the labels have to be 0 or 1 for some of the other methods to work? I don't think this is checked anywhere in the code.

Reply:
it is now!